cleverhans/model_zoo/deep_k_nearest_neighbors/dknn.py

```diff
@@ -0,0 +1,429 @@
+      center = np.mean(self.train_activations_lsh[layer], axis=0)
+      self.train_activations_lsh[layer] -= center
+      self.centers[layer] = center
+
+      # LSH parameters
+      params_cp = falconn.LSHConstructionParameters()
+      params_cp.dimension = len(self.train_activations_lsh[layer][1])
+      params_cp.lsh_family = falconn.LSHFamily.CrossPolytope
+      params_cp.distance_function = falconn.DistanceFunction.EuclideanSquared
+      params_cp.l = self.nb_tables
+      params_cp.num_rotations = 2  # for dense set it to 1; for sparse data set it to 2
+      params_cp.seed = 5721840
+      # we want to use all the available threads to set up
+      params_cp.num_setup_threads = 0
+      params_cp.storage_hash_table = falconn.StorageHashTable.BitPackedFlatHashTable
+
+      # we build 18-bit hashes so that each table has
+      # 2^18 bins; this is a good choice since 2^18 is of the same
+      # order of magnitude as the number of data points
+      falconn.compute_number_of_hash_functions(17, params_cp)
+
+      print('Constructing the LSH table')
+      table = falconn.LSHIndex(params_cp)
+      table.setup(self.train_activations_lsh[layer])
+
+      # Parse test feature vectors and find k nearest neighbors
+      query_object = table.construct_query_object()
+      query_object.set_num_probes(self.nb_tables)
+      self.query_objects[layer] = query_object
+
+  def find_train_knns(self, data_activations):
+    """
+    Given a data_activation dictionary that contains a np array with activations for each layer,
+    find the knns in the training data.
+    """
+    knns_ind = {}
+    knns_labels = {}
+
+    for layer in self.layers:
+      # Pre-process representations of data to normalize and remove training data mean.
+      data_activations_layer = copy.copy(data_activations[layer])
+      nb_data = data_activations_layer.shape[0]
+      data_activations_layer /= np.linalg.norm(
+          data_activations_layer, axis=1).reshape(-1, 1)
+      data_activations_layer -= self.centers[layer]
+
+      # Use FALCONN to find indices of nearest neighbors in training data.
+      knns_ind[layer] = np.zeros(
+          (data_activations_layer.shape[0], self.neighbors), dtype=np.int32)
+      knn_errors = 0
+      for i in range(data_activations_layer.shape[0]):
+        query_res = self.query_objects[layer].find_k_nearest_neighbors(
+            data_activations_layer[i], self.neighbors)
+        try:
+          knns_ind[layer][i, :] = query_res
+        except:  # pylint: disable-msg=W0702
+          knns_ind[layer][i, :len(query_res)] = query_res
+          knn_errors += knns_ind[layer].shape[1] - len(query_res)
+
+      # Find labels of neighbors found in the training data.
+      knns_labels[layer] = np.zeros((nb_data, self.neighbors), dtype=np.int32)
+      for data_id in range(nb_data):
+        knns_labels[layer][data_id,
+                           :] = self.train_labels[knns_ind[layer][data_id]]
```

Comment:
This line-split looks like it's meant for 80 characters. I believe we increased it?

Reply:
Good catch, thanks